azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_mqtt.py

```diff
@@ -133,6 +133,17 @@ def _run_op(self, op):
                 logger.error("transport.reauthorize_connection raised error")
                 logger.error(traceback.format_exc())
                 self._pending_connection_op = None
+                # Send up a DisconenctedEvent.  If we ran a ReauthorizeConnectionOperatoin,
+                # some code must think we're still connected.  If we got an exception here,
+                # we're not conencted, and we need to notify upper layers. (Paho should do this,
+                # but it only causes a DisconnectedEvent on manual disconnect or if a PINGRESP
+                # failed, and it's possible to hit this code without either of those things
+                # happening.
+                # BKTODO: it("sends a DisconnectedEvent up if reauthorize_connection fails")
+                if isinstance(e, transport_exceptions.ConnectionDroppedError) or isinstance(
+                    e, transport_exceptions.ConnectionFailedError
+                ):
+                    self.send_event_up(pipeline_events_base.DisconnectedEvent())
```

Comment:
>self.send_event_up(pipeline_events_base.DisconnectedEvent()) [](start = 20, length = 60)

This is the third bug.  If reconnect/publish/subscribe/unsubscribe fails because of a connection failure, we need to notify upper stages.  Otherwise the pipeline is not connected when it thinks it is.